[AArch64] Enable AvoidLDAPUR for cpu=generic between armv8.4 and armv9.3.

davemgreen opened the pull request:
As added in #124274, CPUs in this range can suffer from performance issues with ldapur. As the gain from ldar->ldapr is expected to be greater than the minor gain from ldapr->ldapur, this opts to avoid the instruction under the default -mcpu=generic when the -march is less that armv8.8 / armv9.3.

I renamed AArch64Subtarget::Others to AArch64Subtarget::Generic to be clearer what it means.

rj-jesus commented:
Hi, thank you for this! I think it looks good overall, though I think we should extend this to Armv8-A too.

Maybe we could replace the condition `hasV9_0aOps() && !hasV9_3aOps()` with `!hasV8_8aOps()`? This should enable `AvoidLDAPUR` prior to v8.8 and (transitively) v9.3. What do you think?

davemgreen commented:
> Hi, thank you for this! I think it looks good overall, though I think we should extend this to Armv8-A too.

Do you expect many people to use armv8.4-a when targeting grace? Or higher armv8 variants of the architecture?

rj-jesus commented:
> Do you expect many people to use armv8.4-a when targeting grace? Or higher armv8 variants of the architecture?

I think that's a reasonable expectation, yes.

davemgreen commented:
I talked to some people internally and they mentioned that some of the the graviton machines are armv8.4, and users could use that to compile for multiple cpu generations. I was hoping to limit the blast radius of this in the long run, but whatever we do will be a bit rubbish and adding 8.4 might be the least bad option.

davemgreen committed:
[AArch64] Enable AvoidLDAPUR for cpu=generic between armv8.4 and armv9.3.

As added in #124274, CPUs in this range can suffer from performance issues with
ldapur. As the gain from ldar->ldapr is expected to be greater than the minor
gain from ldapr->ldapur, this opts to avoid the instruction under the default
-mcpu=generic when the -march is less that armv9.3.

rj-jesus reviewed (approved):
Thank you for making these changes, it looks good now. As you say, I think there isn't a perfect solution here. This seems to be the best compromise.

davemgreen commented:
/cherry-pick 6424abcd6c9c6aa8171c79d0fe0369d3a10da3d5
